cosmoz-grouped-list.js

```diff
@@ -593,6 +590,82 @@
 			if (physicalIndex >= 0) {
 				return this._templateSelectors[physicalIndex].currentElement.__tmplInstance;
 			}
-		}
+		},
+		/**
+		 * Selects/Deselects an item.
+		 * @param {Object} item - The item.
+		 * @param {Boolean} select - True if the item should get selected.
+		 * @returns {undefined}
+		 */
+		_changeItemSelection(item, select) {
+			const model = this._getModelFromItem(item);
+			const setModelState = () => model && (model['selected'] = select);
+
+			if (select && !this.isItemSelected(item)) {
+				this.push('selectedItems', item);
+				setModelState();
+				return;
+			}
+
+			if (select) {
+				return;
+			}
+
+			const idx = this.selectedItems.indexOf(item);
+			if (idx < 0) {
+				return;
+			}
+			this.splice('selectedItems', idx, 1);
+			setModelState();
+		},
+		/**
+		 * Returns the group of an item.
+		 * @param {Object} item - The item.
+		 * @returns {Object} - The group of an item.
+		 */
+		_getGroupOfItem(item) {
+			const groupState = this._groupsMap;
+			if (!groupState) {
+				return;
+			}
+			return this.data.find(group => group.items.find(i => i === item));
+		},
+		/**
+		 * Checks if all items of a group are selected.
+		 * @param {Object} group - The group to make the check on.
+		 * @returns {Boolean} - True if all items of a group are selected.
+		 */
+		_allGroupItemsSelected(group) {
+			if (!group || !group.items) {
+				return;
+			}
+			return group.items.every(item => {
+				const model = this._getModelFromItem(item);
+				return model && model['selected'];
+			});
+		},
+
+		_selectedItemsChanged() {
+			this._debounceValidateGroupSelectionState();
+		},
+
+		_debounceValidateGroupSelectionState() {
+			this.debounce('_validateGroupSelectionState', this._validateGroupSelectionState, 10);
+		},
+		/**
+		 * Selects a group if all items of a group are selected.
+		 * Deselects a group if not all items of a group are selected.
+		 * @returns {undefined}
+		 */
+		_validateGroupSelectionState() {
+			const groupState = this._groupsMap;
+			if (!groupState) {
+				return;
+			}
+			this.selectedItems.forEach(item => {
```

Comment:
Can't this be optimized a bit ?
For each selected items, we search for its group and check if all items in the group are selected.
Wouldn't it be more efficient for example to get distinct groups for selected items and then check if all items are selected ?